torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1314,6 +1317,24 @@ void ProcessGroupNCCL::workCleanupLoop() {
     // Bump up heart beat by one.
     heartbeat_++;
 
+    // poll store to see if some ranks have flagged a timeout when
+    // we haven't polled for `heartbeat_timeout` seconds and there haven't
+    // any work added or removed for `watchdog_timeout` seconds.
+    if (dumpOnTimeout_) {
+      auto currentTimepoint = std::chrono::steady_clock::now();
+      if (std::chrono::duration_cast<std::chrono::milliseconds>(
+              currentTimepoint - lastWorkListUpdateTime_) >=
+              std::chrono::milliseconds(kWatchdogThreadSleepMillis) &&
+          std::chrono::duration_cast<std::chrono::seconds>(
+              currentTimepoint - lastTimePollStore) >=
+              std::chrono::seconds(heartbeatTimeoutInSec_)) {
+        lastTimePollStore = currentTimepoint;
+        if (store_->check({std::string(TIMEOUT_DUMP)}) && !optAsyncDebugDump) {
+          optAsyncDebugDump = launchAsyncDebugDump();
+        }
```

Comment:
should we kill the process here? we know someone else timed out

Reply:
yes, will add a wait_for and throw an exception here.